Fixed #26749 -- Preserved behavior of use_for_related_field during deprecation.

spectras opened the pull request:
Follow-up of #6825.

* Updates test cases
* Adjusts deprecation warning texts
* Adds corner case addressed by #6825, providing a workaround for [ticket 26749](https://code.djangoproject.com/ticket/26749)

As it's intended to be a temporary workaround for v1.10, I based the change on top of `stable.1.10.x`, but I can move it back to `master` if needed. I can also merge the two commits.

spectras commented:
Right, flake8… beware, I amended my commit to add a few blank lines.
In case you hit a conflict, overwrite my changes I'll do them again.

timgraham commented:
buildbot, add to whitelist.

timgraham commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -110,7 +110,8 @@ def get_queryset(self, **hints):
                 warnings.warn(
                     "use_for_related_fields is deprecated, instead "
-                    "set Meta.base_manager_name on '{}'.".format(related_model._meta.label),
+                    "set Meta.base_manager_name on '{}' then remove "
```
I don't think "then remove use_for_related_fields" is needed. Removing usage of the deprecated feature is always implied by deprecation warnings. Can the two options not be used together though? That might be problematic for projects that want to support multiple versions of Django.

spectras replied:
They can. The exact semantics are as follows:

* Legacy way alone => old behavior, deprecation warning.
* Mixing legacy and new way => mixed behavior, deprecation warning. *This is the point of this PR, as new way is not fully backwards compatible with old way and updating the API is too late for release 1.10*.
* New way alone => new behavior, no warning

I thought the warning would allow expliciting that new behavior is not fully active until `use_for_related_fields` is not removed. You have the final word anyway, if you prefer that I drop the "then remove use_for_related_fields", I can do so.

timgraham commented on tests/managers_regress/tests.py:
```diff
@@ -375,16 +391,61 @@ class MyModel2(models.Model):
                 pass
         self.assertEqual(len(warns), 0)
 
+        # When mixing the new base_manager_name API and use_for_related_fields, there should be warnings.
```
Please wrap inline comments at 79 characters.

timgraham commented:
Yes, for future reference, please send the pull request against master and let the committer backport it. Since there aren't likely to be conflicts it doesn't matter too much. We can cherry-pick to merge it.

spectras commented:
Alright, I need your input on the warning message and I'll improve the change.

=> Do you want the commits to be squashed? And perhaps rebased on master?

timgraham commented:
I'd add the semantics to the release notes. No strong opinion about the warning message, I guess. "instead" implies removing use_for_related_fields to me.

Squashing is fine. Don't need to rebase on master since that would require opening a new PR.

spectras commented:
Squashed, and added a note for the behavior into the release notes.
Now it's described there, I feel comfortable with reverting the warning message, and I did so.

timgraham commented:
merged in f4afb85d7e900123fa8f88110adc011ab184e153, thanks!
